javalanglib/src/main/scala/java/lang/ObjectClone.scala

```diff
@@ -0,0 +1,85 @@
+package java.lang
+
+import scala.scalajs.js
+
+/** Implementation of `java.lang.Object.clone()`.
+ *
+ *  Called by the hard-coded IR of `java.lang.Object`.
+ */
+private[lang] object ObjectClone {
+  private val getOwnPropertyDescriptors: js.Function1[js.Object, js.Object] = {
+    import js.Dynamic.{global, literal}
+
+    // The val f = ...; f.asInstanceOf's are necessary for 2.10 not to crash
+
+    // Fetch or polyfill Object.getOwnPropertyDescriptors
+    if (js.typeOf(global.Object.getOwnPropertyDescriptors) == "function") {
+      val f = global.Object.getOwnPropertyDescriptors
+      f.asInstanceOf[js.Function1[js.Object, js.Object]]
+    } else {
+      // Fetch or polyfill Reflect.ownKeys
+      type OwnKeysType = js.Function1[js.Object, js.Array[js.Any]]
+      val ownKeysFun: OwnKeysType = {
+        if (js.typeOf(global.Reflect) != "undefined" &&
+            js.typeOf(global.Reflect.ownKeys) == "function") {
+          val f = global.Reflect.ownKeys
+          f.asInstanceOf[OwnKeysType]
+        } else {
+          // Fetch Object.getOwnPropertyNames
+          val getOwnPropertyNames = {
+            val f = global.Object.getOwnPropertyNames
+            f.asInstanceOf[OwnKeysType]
+          }
+
+          // Fetch or polyfill Object.getOwnPropertySymbols
+          val getOwnPropertySymbols: OwnKeysType = {
+            if (js.typeOf(global.Object.getOwnPropertySymbols) == "function") {
+              val f = global.Object.getOwnPropertySymbols
+              f.asInstanceOf[OwnKeysType]
+            } else {
+              /* Polyfill for Object.getOwnPropertySymbols.
+               * We assume that if that function does not exist, then symbols
+               * do not exist at all. Therefore, the result is always an empty
+               * array.
+               */
+              { (o: js.Object) =>
+                js.Array[js.Any]()
+              }
+            }
+          }
+
+          // Polyfill for Reflect.ownKeys
+          { (o: js.Object) =>
+            getOwnPropertyNames(o) ++ getOwnPropertySymbols(o)
+          }
+        }
+      }
+
+      // Polyfill for Object.getOwnPropertyDescriptors
+      { (o: js.Object) =>
+        val ownKeys = ownKeysFun(o)
+        val descriptors = new js.Object
+        for (key <- ownKeys) {
+          global.Object.defineProperty(descriptors, key, literal(
+              configurable = true,
+              enumerable = true,
+              writable = true,
+              value = global.Object.getOwnPropertyDescriptor(o, key)
+          ))
+        }
```

Comment:
How is this different than just assigning the property descriptor to `descriptors`?

Reply:
It is equivalent, *except* if there is an accessor property named `key` in `o` or its prototype chain. If that's the case, the assignment calls the setter, whereas `defineProperty` by-passes the (inherited) setter and creates a data property on `o`.

Since `descriptors` is a brand `new js.Object`, this can only happen if there is an accessor property on `Object.prototype`, which is unlikely but possible.

Reply:
Ah, fair enough. Maybe add a comment to explain this? It's kind of non-obivous :)